cpp/src/db/db_object.cuh

```diff
@@ -20,172 +20,178 @@
 #include <vector>
 #include <map>
 #include "utilities/graph_utils.cuh"
+#include "rmm/device_buffer.hpp"
 
-namespace cugraph { 
+namespace cugraph {
 namespace db {
-  /**
-   * Class for representing an entry in a pattern, which may either be a variable or constant value
-   * See description of db_pattern for more info on how this is used.
-   */
-  template <typename idx_t>
-  class db_pattern_entry {
-    bool is_var;
-    idx_t constantValue;
-    std::string variableName;
+/**
+ * Class for representing an entry in a pattern, which may either be a variable or constant value
+ * See description of db_pattern for more info on how this is used.
+ */
+template<typename idx_t>
+class db_pattern_entry {
+  bool is_var;
+  idx_t constantValue;
+  std::string variableName;
+  public:
+  db_pattern_entry(std::string variable);
+  db_pattern_entry(idx_t constant);
+  db_pattern_entry(const db_pattern_entry<idx_t>& other);
+  db_pattern_entry& operator=(const db_pattern_entry<idx_t>& other);
+  bool isVariable() const;
+  idx_t getConstant() const;
+  std::string getVariable() const;
+};
+
+/**
+ * Class for representing a pattern (usually a triple pattern, but it's extensible)
+ * A pattern in this sense consists of a sequence of entries each element is either a constant
+ * value (an integer, since we dictionary encode everything) or a variable. Variables stand
+ * in for unknown values that are being searched for. For example: if we have a pattern like
+ * {'a', :haslabel, Person} (Where :haslabel and Person are dictionary encoded constants and
+ * 'a' is a variable) We are looking for all nodes that have the label Person.
+ */
+template<typename idx_t>
+class db_pattern {
+  std::vector<db_pattern_entry<idx_t>> entries;
   public:
-    db_pattern_entry(std::string variable);
-    db_pattern_entry(idx_t constant);
-    db_pattern_entry(const db_pattern_entry<idx_t>& other);
-    db_pattern_entry& operator=(const db_pattern_entry<idx_t>& other);
-    bool isVariable() const;
-    idx_t getConstant() const;
-    std::string getVariable() const;
-  };
+  db_pattern();
+  db_pattern(const db_pattern<idx_t>& other);
+  db_pattern& operator=(const db_pattern<idx_t>& other);
+  int getSize() const;
+  const db_pattern_entry<idx_t>& getEntry(int position) const;
+  void addEntry(db_pattern_entry<idx_t>& entry);
+  bool isAllConstants();
+};
+
+/**
+ * Class which encapsulates a CSR-style index on a column
+ */
+template<typename idx_t>
+class db_column_index {
+  rmm::device_buffer offsets;
+  rmm::device_buffer indirection;
+  idx_t offsets_size;
+  idx_t indirection_size;
```

Comment:
```suggestion
  idx_t offsets_size{0};
  idx_t indirection_size{0};
```

Reply:
And if you initialize here, no need to define your own default constructor, destructor, copy & move constructor & operator.

Just mark them to use default one (e.g. db_column_index() = default;) will work and will be much easier to understand.